### ISSUE TITLE
[tester] add an option for parallel test execution

### ISSUE THREAD

**Issue by laeubi** [bndtools/bnd#5509]
I'm currently running the CM-TCK with https://bnd.bndtools.org/chapters/310-testing.html and it has many tests and is very slow... still most of my CPUs are idle. 

It would be good to have an option to execute tests in parallel e.g. `tester.threads=<a number>`

**Comment by bjhargrave** [comment]
This sounds like an interesting addition. 

But I don't think this would help with the CM tests since they manipulate a database (CM configurations) and then confirm the state of the database through events and inspection, and running multiple tests at the same time may cause conflicts. To run the CM TCK tests in parallel, each test would need its own isolated framework and CM instance.





> is very slow... still most of my CPUs are idle.

Also, this slowness is about waits in the test for negative testing. For example, do some action and test that something does not happen (like an event delivery). These tests use waits before confirming the something did not happen. Not perfect but it is what it is. The CM TCK tests could use a thorough update. See some of the draft PRs in the osgi repo where some changes are proposed.

**Comment by laeubi** [comment]
> But I don't think this would help with the CM tests since they manipulate a database (CM configurations) and then confirm the state of the database through events and inspection, and running multiple tests at the same time may cause conflicts.

Okay this might then not work for CM but probably for others so still useful in the long run if we can ever make CM test parallel, currently they take 17 minutes and waiting most of the time :-)

> To run the CM TCK tests in parallel, each test would need its own isolated framework and CM instance.

I'll open a separate issue for that.

**Comment by kriegfrj** [comment]
@laeubi, it has been a while, but I think it is already possible to run tests in parallel. I have video evidence of having done it before (see https://github.com/bndtools/bnd/pull/4310). At least, it should be possible for tests using the Jupiter engine (not sure about Vintage/JUnit 4 tests).

It has been a while and the config for my little test case has long since been consigned to the digital dustbin, however I think that this simply requires setting the right Jupiter property `junit.jupiter.execution.parallel.enabled` to `true` using `-runproperties` or something similar. See [the JUnit 5 docs](https://junit.org/junit5/docs/snapshot/user-guide/#writing-tests-parallel-execution) for more info.

Of course, you may run into concurrency issues as @bjhargrave noted, so a fork option would be nice. Otherwise, judicious use of the resource locking feature might help, but it might also cause the entire suite to degenerate into an effective single-threaded execution.

**Comment by kriegfrj** [comment]
> @laeubi, it has been a while, but I think it is already possible to run tests in parallel. I have video evidence of having done it before (see #4310). At least, it should be possible for tests using the Jupiter engine (not sure about Vintage/JUnit 4 tests).
> 
> It has been a while and the config for my little test case has long since been consigned to the digital dustbin, however I think that this simply requires setting the right Jupiter property `junit.jupiter.execution.parallel.enabled` to `true` using `-runproperties` or something similar. See [the JUnit 5 docs](https://junit.org/junit5/docs/snapshot/user-guide/#writing-tests-parallel-execution) for more info.

Btw, if you do manage to get this to work, let us know so that we can mark the issue as resolved.

**Comment by laeubi** [comment]
> Btw, if you do manage to get this to work, let us know so that we can mark the issue as resolved.

If you can add that info to the tester-page I think that would resolve the issue of "running in parallel" and for the isolation I think 
- https://github.com/bndtools/bnd/issues/5510

is already a follow up.

**Comment by pkriens** [comment]
Gradle already supports testing in parallel? In the GUI it would be confusing imho. 

I'd like to close issues without someone that creates a PR. So will you?
